### ISSUE TITLE
Count the reads at each processing step for each sample

### ISSUE THREAD

**Issue by ewallace** [riboviz/RiboViz#108]
As part of quality control, we need to summarize how many reads are processed at each stage. For example, how many fastq reads are present in the initial file, how many aligned to rRNA, etc. Currently this information is in two places:

* Log files for individual steps, in very different formats (probably faster to run)
* In the numbers of lines of the .fastq/.sam/.bam files, most of which are in the `tmp` folder

We would like to summarize this information per-sample, perhaps in a module of the successor .Rmd file to `generate_stats_figs.R` mentioned in #32.

**Comment by mikej888** [comment]
Depends on https://github.com/riboviz/RiboViz/issues/32#issuecomment-567970484 - perhaps better to wait until that is resolved but identification of how to extract the reads can be done in the meantime.

**Comment by ewallace** [comment]
I don't think this depends on #32, because it is only a question of counting reads in the .fastq/.bam/.sam files, all of which are created upstream of `bam_to_h5.R`.

**Comment by mikej888** [comment]
Options for extracting data

### Inputs

* Count records in `fq` file.
* This could be optional since `cutadapt` outputs number of reads input.

Suggest:

* Rely on `cutadapt` to avoid need to initially traverse `fq` file.

### `cutadapt`

* Parse log file:

```
Total reads processed:                 <N>
Reads written (passing filters):       <N> (<P>%)
```

* Parse `--report=minimal` TSV log file:
  - `in_reads`
  - `out_reads`
* Count records in `fq` file.

Suggest:

* Parse `--report=minimal` TSV log file, since it is machine-readable, unlike log file, and avoids need to scan `fq` file.

### `umi_tools extract`

* Parse log file:

```
YYYY-MM-DD HH:MM:SS,MMM INFO Input Reads: <N>
YYYY-MM-DD HH:MM:SS,MMM INFO Reads output: <N>
```

* Parse `-L FILE`:

```
YYYY-MM-DD HH:MM:SS,MMM INFO Input Reads: <N>
YYYY-MM-DD HH:MM:SS,MMM INFO Reads output: <N>
```

* Count records in `fq` file.
* Ignore as no reads are removed.

Suggest:

* Ignore as no reads are removed.

### `riboviz.tools.demultiplex_fastq`

* Parse log file:

```
All <N> reads processed
```

* Parse `-o DIRECTORY/num_reads.tsv` TSV file:
  - `SampleID` - for each sample, `Unassigned`, `Total`
  - `NumReads`
* Count records in sample-specific and `Unassigned.fastq` files.

Suggest:

* Parse `DIRECTORY/num_reads.tsv` TSV file, since it is machine-readable, unlike log file, and avoids need to scan multiple `fq` files.

### `hisat2`

Processed reads:

* Parse log file:

```
<N> reads; of these:
  <N> (<P>%) were unpaired; of these:
```

* Parse `--summary-file` file: as for log file above.
* Parse log file with `--new-summary` flag:

```
	Total reads: <N>
```

Unaligned reads:

* Parse log file:

```
    <N> (<P>%) aligned 0 times
```

* Parse `--summary-file` file: as for log file above.
* Parse log file with `--new-summary` flag:

```
		Aligned 0 time: <N> (<P>%)
```

* Count records in `fq` file.

Aligned reads:

* Parse log file:

```
    <N> (<P>%) aligned exactly 1 time
    <N> (<P>%) aligned >1 times
```

* Parse `--summary-file` file: as for log file above.
* Parse log file with `--new-summary` flag:

```
		Aligned 1 time: <N> (<P>%)
		Aligned >1 times: <N> (<P>%)
```

* Count records in `sam` file (`-c -F 260`) for number of reads value consistent with output from hisat2 (i.e. count of mapped (primary aligned) reads only).
* Count records in `sam` file (`-c`) for number of reads value consistent with downstream processing of `sam` file (i.e. count of all reads).
* Count records in `sam` file using custom script to output both values above with only one pass through the `sam` file.

Suggest:

* Parse `--new-summary` log file.
* Count records in `sam` file using custom script to output both count of mapped (primary aligned) reads and count of all reads.

### `riboviz.tools.trim_5p_mismatch`

* Parse log file:

```
processed:	<N>
written:	<N>
```

* Count records in `sam` file (`-c`).
* Extend to write machine-readable TSV file:
  - `processed`
  - `discarded`
  - `trimmed`
  - `written`

Suggest:

* Extend to write machine-readable TSV file and then parse this file to avoid need to scan `sam` file.

### `umi_tools dedup`

* Parse log file:

```
YYYY-MM-DD HH:MM:SS,MMM INFO Reads: Input Reads: <N>
YYYY-MM-DD HH:MM:SS,MMM INFO Number of reads out: <N>
```

* Parse `-L FILE`:

```
YYYY-MM-DD HH:MM:SS,MMM INFO Reads: Input Reads: <N>
YYYY-MM-DD HH:MM:SS,MMM INFO Number of reads out: <N>
```

* Count records in `bam` file (`-c`).
* Count records in `umi_tools group` `--group-out` TSV file, post-deduplication. But, this is an optional step.

Suggest:

* Parse `-L FILE` log file to avoid the need to scan the `bam` file.

**Comment by mikej888** [comment]
Possible summary file format

```
Stage  Sample        Type                File        Reads
<STEP> <SAMPLE>|NULL NULL|in|out|discard <FILE>|NULL <NUMBER_OF_READS>
```

`vignette_config.yaml` example:

```
Stage            Sample Type    File                     Reads
input            WTnone         SRR1042855_s1mi.fastq.gz  963571
cutadapt         WTnone in      SRR1042855_s1mi.fastq.gz  963571
cutadapt         WTnone out     trim.fq                   952343
cutadapt         WTnone discard                            11228
hisat2 (1)       WTnone in      trim.fq                   952343
hisat2 (1)       WTnone out     nonrRNA.fq                466464
hisat2 (1)       WTnone out     rRNA_map.sam              485879 
hisat2 (1)       WTnone out     rRNA_map.sam             1430213 # mapped (primary aligned) reads
hisat2 (2)       WTnone in      nonrRNA.fq                446464
hisat2 (2)       WTnone out     unaligned.fq              452266
hisat2 (2)       WTnone out     orf_map.sam                14198
hisat2 (2)       WTnone out     orf_map.sam                14516 # mapped (primary aligned) reads
trim_5p_mismatch WTnone in      orf_map.sam                14516
trim_5p_mismatch WTnone out     orf_map_clean.sam          14516
input            WT3AT          SRR1042864_s1mi.fastq.gz 1374448
cutadapt         WT3AT  in      SRR1042864_s1mi.fastq.gz 1374448
cutadapt         WT3AT  out     trim.fq                  1373362
cutadapt         WT3AT  discard                             1086
hisat2 (1)       WT3AT  in      trim.fq                  1373362
hisat2 (1)       WT3AT  out     nonrRNA.fq                485226
hisat2 (1)       WT3AT  out     rRNA_map.sam              888136
hisat2 (1)       WT3AT  out     rRNA_map.sam             2254078 # mapped (primary aligned) reads
hisat2 (2)       WT3AT  in      nonrRNA.fq                485226
hisat2 (2)       WT3AT  out     unaligned.fq              452266
hisat2 (2)       WT3AT  out     orf_map.sam                 8441
hisat2 (2)       WT3AT  out     orf_map.sam                 8698 # mapped (primary aligned) reads
trim_5p_mismatch WT3AT  in      orf_map.sam                 8698
trim_5p_mismatch WT3AT  out     orf_map_clean.sam           8698
```

`simdata_umi_config.yaml` excerpt:

```
Stage            Sample    Type    File                        Reads
input            umi5_umi3 -       umi5_umi3_umi_adaptor.fastq 9
cutadapt         umi5_umi3 in      umi5_umi3_umi_adaptor.fastq 9
cutadapt         umi5_umi3 out     trim.fq                     9
cutadapt         umi5_umi3 discard                             0
hisat2 (1)       umi5_umi3 in      trim.fq                     9
hisat2 (1)       umi5_umi3 out     nonrRNA.fq                  9
hisat2 (1)       umi5_umi3 out     rRNA_map.sam                0
hisat2 (1)       umi5_umi3 out     rRNA_map.sam                9 # mapped (primary aligned) reads
hisat2 (2)       umi5_umi3 in      nonrRNA.fq                  9
hisat2 (2)       umi5_umi3 out     unaligned.fq                0
hisat2 (2)       umi5_umi3 out     orf_map.sam                 9
hisat2 (2)       umi5_umi3 out     orf_map.sam                 9 # mapped (primary aligned) reads
trim_5p_mismatch umi5_umi3 in      orf_map.sam                 9
trim_5p_mismatch umi5_umi3 out     orf_map_clean.sam           9
umi_tools dedup  umi5_umi3 in      pre_dedup.bam               9
umi_tools dedup  umi5_umi3 out     umi5_umi3.bam               5
```

`simdata_multiplex_config.yaml` excerpt:

```
Stage             Sample Type File                                          Reads
input                         multiplex_umi_barcode_adaptor.fastq           90
cutadapt                 in   multiplex_umi_barcode_adaptor.fastq           90
cutadapt                 in   multiplex_umi_barcode_adaptor.fastq           90
cutadapt                 out  multiplex_umi_barcode_adaptor_trim.fq         90
demultiplex_fastq        in   multiplex_umi_barcode_adaptor_extract_trim.fq 90
demultiplex_fastq Tag0   out  Tag0.fastq                                    27
demultiplex_fastq Tag1   out  Tag1.fastq                                    27
demultiplex_fastq Tag2   out  Tag2.fastq                                    27
demultiplex_fastq        out  Unassigned.fastq                               9
```

Questions:

* How to note distinction between number of mapped (primary aligned) reads in SAM file (whose value is output by `hisat2`) and total number of reads in SAM file whose value is output by `trim_5p_mismatch`?
* Omit Type column and only information about record number of reads in files output at each step. For example:

```
Stage            Sample File                     Reads
input            WTnone SRR1042855_s1mi.fastq.gz  963571
cutadapt         WTnone SRR1042855_s1mi.fastq.gz  963571
cutadapt         WTnone trim.fq                   952343
cutadapt         WTnone discard                    11228
hisat2 (1)       WTnone nonrRNA.fq                466464
hisat2 (1)       WTnone rRNA_map.sam              485879 
hisat2 (1)       WTnone rRNA_map.sam             1430213 # mapped (primary aligned) reads
hisat2 (2)       WTnone unaligned.fq              452266
hisat2 (2)       WTnone orf_map.sam                14198
hisat2 (2)       WTnone orf_map.sam                14516 # mapped (primary aligned) reads
trim_5p_mismatch WTnone orf_map_clean.sam          14516
```

**Comment by mikej888** [comment]
Decisions at 15/01/20 meeting:

* Output TSV file should be of form:

```
SampleName | Human summary | Program | File | Count |
```

* Only the number of output reads needs to be included.
* Include a provenance header comment.

* Python script to build summary file should post-process output files and should be the final operation called by prep_riboviz.py, controlled by a configuration parameter. A complementary script can allow this to be run standalone.

**Comment by mikej888** [comment]
[count-reads-108](https://github.com/riboviz/RiboViz/tree/count-reads-108) bransh as of commit  367a680 has `riboviz.workflow_record` module with functions to create and update a TSV file with columns:

* SampleName
* Description
* Program
* File
* Read/Write: "read" or "write"

For example:
```
# Date: 2020-01-21 10:05:28.455905
# Command-line tool: /home/ubuntu/riboviz/riboviz/tools/prep_riboviz.py
# File: /home/ubuntu/riboviz/riboviz/workflow_record.py
# Version: commit 367a680d08e8528fc1a46eddaeea9c52c58b13bb date 2020-01-21 10:05
:20-08:00
SampleName	Description	Program	File	Read/Write
	Build indices for alignment	hisat2-build	vignette/input/yeast_rRNA_R64-1-1.fa	read
	Build indices for alignment	hisat2-build	vignette/index/yeast_rRNA*	write
	Build indices for alignment	hisat2-build	vignette/input/yeast_YAL_CDS_w_250utrs.fa	read
	Build indices for alignment	hisat2-build	vignette/index/YAL_CDS_w_250*	write
	Cut out sequencing library adapters	cutadapt	data/simdata/multiplex_umi_barcode_adaptor.fastq	read
	Cut out sequencing library adapters	cutadapt	vignette/simdata_multiplex_tmp/multiplex_umi_barcode_adaptor_trim.fq	write
	Extract barcodes and UMIs	umitools extract	vignette/simdata_multiplex_tmp/multiplex_umi_barcode_adaptor_trim.fq	read
...
Tag0	Remove rRNA or other contaminating reads by alignment to rRNA index files	hisat2	vignette/simdata_multiplex_tmp/multiplex_umi_barcode_adaptor_deplex/Tag0.fastq	read
Tag0	Remove rRNA or other contaminating reads by alignment to rRNA index files	hisat2	vignette/index/yeast_rRNA*	read
Tag0	Remove rRNA or other contaminating reads by alignment to rRNA index files	hisat2	vignette/simdata_multiplex_tmp/Tag0/nonrRNA.fq	write
Tag0	Remove rRNA or other contaminating reads by alignment to rRNA index files	hisat2	vignette/simdata_multiplex_tmp/Tag0/rRNA_map.sam	write
...
```

This records the files and directories produced at each step. Building it as the workflow executes reduces the need to re-infer what temporary and output files were produced at each step. This TSV can be post-processed to:

* Identify the files for which read counts need to be calculated.
* In future, to help produce a .dot file visualising the exact workflow that was run.

**Comment by mikej888** [comment]
As of commit 3c1e660 all files produced are logged - including individual index files and contents of directories with demultiplexed files.

**Comment by mikej888** [comment]
To commit 9bff035 completes logging of files read and written by workflow steps.

Functions in `riboviz.workflow` invoke third-party commands. These have been extended to invoke `riboviz.workflow_files_logger.log_files`. Different approaches are used to determine the files to log depending on the specific command being invoked.

* Input and output filenames provided as arguments to these functions in `riboviz.workflow` are recorded as-is.
* For file prefixes (e.g. index file prefixes read/written by `hisat2-build`, `hisat2`, `umi_tools dedup` statistics files), all files matching the prefix (found via `glob` on the prefix) are recorded.
* For `.bam.bai` files, `.bam.bai` file names recorded are derived from the `.bam` file (using a format string in `riboviz.sam_bam`).
* For `*.tsv` and `*.pdf` files written by `generate_stats_figs`, all files with these extensions in the output directory (found via `glob` on these patterns) are recorded.
* For sample-specific `tpms.tsv` files read by `collate_tpms.R`, `<OUTPUT>/<SAMPLE>/tpms.tsv` files (where `tpms.tsv` is defined in `riboviz.workflow_r`) are recorded.
* `<OUTPUT>/TPMs_collated.tsv` written by `collate_tpms.R` is recorded (where `TPMS_collated.tsv` is defined as a constant in `riboviz.workflow_r`).
* For the contents of the demultiplexed files directory, output by `riboviz.tools.demultiplex_fastq`, all the files in this directory (found via `glob`) are recorded.

A `riboviz.workflow_files_logger.validate_log_file` function is used by `riboviz.tools.prep_riboviz` to check that all files recorded in workflow record file exist and that all files in a list of directories are recorded. Any files that are missing or are unrecorded are logged with a warning.

**Comment by mikej888** [comment]
For counting reads, the workflow files log file can be post-processed to filter all rows whose `Read/Write` value is `write` then filtering on extensions of the `File` value to identify the files with read-specific information.

**Comment by mikej888** [comment]
Remaining tasks are to identify and log all input files as "input" then write modules to count reads.

**Comment by mikej888** [comment]
* Updated `riboviz.tools.prep_riboviz` to check that sample-independent input files exists before processing of samples begins to avoid repeated existence checks when every sample is processed.
* Updated `riboviz.tools.prep_riboviz` to log input files in workflow files log file.
* Added `count_sequences` functions to both `riboviz.sam_bam` and `riboviz.fastq` to count sequences in FASTQ, SAM and BAM files. Added associated tests to `riboviz.test.test_sam_bam` and `riboviz.test.test_fastq`.
* Working on `riboviz.tools.count_reads` to parse workflow files log file and count reads. Currently prints information on reads counted to stdout.

**Comment by mikej888** [comment]
[count-reads-108](https://github.com/riboviz/RiboViz/tree/count-reads-108) branch:

* Uses a workflow files logs file.
* Could be used to create DOT file in future.
* count_reads parses workflow files logs file, creates read_counts.tsv, then logs this into the logs file.
* count_reads uses Pandas to query workflow files logs file but it feels overengineered.
* Can't use count_reads within the bash script, as that can't create the workflow files logs file.

[count-reads-108-alt](https://github.com/riboviz/RiboViz/tree/count-reads-108-alt) branch:

* Branch of the above.
* count_reads no longer parses workflow files logs file. It crawls directories, creates read_counts.tsv, then logs this into the logs file.
* Can use count_reads from within bash script.
* Mapping from input files to sample IDs is lost - how to capture for use by count_reads and how would bash script capture this? A lightweight "register_sample.py <SAMPLE_ID> <FILE>" script that creates a "SampleID|File" TSV file, used by the workflow and added to the bash script.

**Comment by mikej888** [comment]
Discussed at developer meeting of 07/02/2020. We agreed to use the count-reads-108-alt solution. The mapping from input files to sample names can be provided by using the YAML configuration file. It was also agreed that the workflow files logs file-related code would be removed as it is no longer needed (if needed it can be retrieved from Git).

**Comment by mikej888** [comment]
Merged via pull requests #141, #142
